### PR TITLE
[BEAM-6935] Spark portable runner: implement side inputs

### PR DESCRIPTION
**Pull request by ibzib**
1. During translation, convert each side input's corresponding RDD into bytes, which bytes are then broadcast to the Spark context. (This is necessary because in Spark we cannot broadcast an RDD directly like how we can broadcast a Dataset in Flink.)
2. Pass reference to the resulting Spark broadcast variable along with a decoder for those bytes to the executable stage function.
3. When the executable stage function needs to read the side inputs, it gets the broadcasted bytes value, then uses the decoder to convert the bytes to the side input value.

Notes:
- Converted `FlinkBatchSideInputHandlerFactory` to a pluggable `BatchSideInputHandlerFactory` to connect Spark implementation with portability framework.
- For now, testing is still just a log statement with the side input value. More formal verification is still a few PRs away.
- It is quite possible we will want to improve performance in the future; this is just a starting point.

R: @angoenka @iemejia 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java [hunk 1]
```diff
@@ -69,6 +70,14 @@
     return rdd;
   }
 
+  List<byte[]> getBytes(WindowedValue.WindowedValueCoder<T> wvCoder) {
+    if (clientBytes == null) {
+      JavaRDDLike<byte[], ?> bytesRDD = rdd.map(CoderHelpers.toByteFunction(wvCoder));
+      clientBytes = bytesRDD.collect();
```
Comment (angoenka) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java]: Is bytesRDD.collect expected to be invoked only once?

Comment (ibzib) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java]: `getBytes` implicitly assumes a 1:1 relationship between `this` and the `wvCoder` argument, meaning the value of `bytesRDD.collect` would always be the same for each BoundedDataset object and therefore would be cacheable. (This is similar for the existing `getValues` function.)

Comment (angoenka) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java]: I would ideally make these method stateless but I think its fine for now as its the same in getValue

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by angoenka** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
When running this on Python, I'm running into some issues with Spark broadcast variables. When I try to get the value of a broadcast variable, a NullPointerException is thrown. This probably has something to do with the way Spark passes broadcast variables via closures. I will need to update this PR when I have a solution.

**Comment by ibzib** [comment]
> When running this on Python, I'm running into some issues with Spark broadcast variables.

False alarm, there was a different underlying issue. It seems to be working fine.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java [hunk 2]
```diff
@@ -46,6 +46,7 @@
   private Iterable<WindowedValue<T>> windowedValues;
   private Coder<T> coder;
   private JavaRDD<WindowedValue<T>> rdd;
+  private List<byte[]> clientBytes;
```
Comment (iemejia) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java]: kinda have issues with that name but cannot think at the moment on something better apart of the even worse `bytes`

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java [hunk 3]
```diff
@@ -191,6 +209,29 @@ private static void translateImpulse(
...
+    BoundedDataset<T> dataset = (BoundedDataset<T>) context.popDataset(collectionId);
+    PCollectionNode collectionNode = PipelineNode.pCollection(collectionId, collection);
+    WindowedValueCoder<T> coder;
+    try {
+      coder =
+          (WindowedValueCoder<T>) WireCoders.instantiateRunnerWireCoder(collectionNode, components);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    List<byte[]> bytes = dataset.getBytes(coder);
```
Comment (iemejia) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java]: again uneasy about naming but without real propositions, maybe `dataAsBytes` (mmm nothing seems good enough).

**Comment by iemejia** [comment]
This means that we can have now WordCount in place, have you already tried it? Next probably will need to document a bit the user side + website (I can help with that).

**Comment by iemejia** [comment]
As with the other one I let you deal with the JIRA, not sure if you have a different jira for the streaming case.

**Comment by ibzib** [comment]
> This means that we can have now WordCount in place, have you already tried it?

I have tried it and it works, but we still have to merge a few more PRs before it's ready to put on the website.

**Pull request merged**